### PR TITLE
fix(TypographyOverflow): improve performance

### PR DESCRIPTION
**Pull request by s0ber**
[SPC-1148]

### Description

TypographyOverflow is implemented in a way that it always spawns the new style tag whenever new component is rendered. E.g. on the talent list page more than 1000 style tags were spawned. It turned out those style tags were causing style recalculations and unnecessary reflows (and generally they are pretty expensive).

This was happening because we were using the dynamic styles (https://cssinjs.org/jss-plugin-rule-value-function/?v=v10.7.1). This plugin works in a way that it spawns a new stylesheet which holds all dynamic rules for any particular instance of the component. And then updates it with dynamic values when props are changing:
https://github.com/mui-org/material-ui/blob/next/packages/material-ui-styles/src/makeStyles/makeStyles.js#L107
https://github.com/mui-org/material-ui/blob/next/packages/material-ui-styles/src/makeStyles/makeStyles.js#L131

In our case we're not really changing the number of TypographyOverflow lines dynamically. We just use some initial lines value and that is enough.

In this PR I've changed the implementation to use inline styles for `-webkit-line-clamp` property since it is used very rarely and only for multiline typography overflows.

### How to test

- Make sure TypographyOverflow works as expected

### Screenshots

**Before**
Снимок экрана 2021-07-06 в 0 08 57
Снимок экрана 2021-07-06 в 0 14 02

**After**
Снимок экрана 2021-07-06 в 13 25 40

### Search form rendering on talents page
**Without the fix**
Снимок экрана 2021-07-06 в 0 28 50
Снимок экрана 2021-07-06 в 0 28 55

**With the fix**
Снимок экрана 2021-07-06 в 0 26 56
Снимок экрана 2021-07-06 в 0 26 59

### Single talent card rendering

**Without the fix**
Снимок экрана 2021-07-06 в 0 31 49
Снимок экрана 2021-07-06 в 0 31 52

**With the fix**
Снимок экрана 2021-07-06 в 0 32 03
Снимок экрана 2021-07-06 в 0 32 08

This fix affects the performance of almost all pages in SP project and cuts off seconds of UI freezes from some interactions. E.g. the talent list rendering is 5 seconds faster.

### Review

- [x] Read [CONTRIBUTING.md](https://github.com/toptal/picasso/blob/master/CONTRIBUTING.md) and [Component API principles](https://github.com/toptal/picasso/blob/master/docs/api-principles.md)
- [x] Annotate all `props` in component with documentation
- [x] Create `examples` for component
- [x] Ensure that deployed demo has expected results and good examples
- [x] Ensure that tests pass by running `yarn test`
- [x] Ensure that visuals tests pass by running `yarn test:visual`. If not - check the documentation [how to fix visual tests](https://github.com/toptal/picasso/blob/master/docs/contribution/visual-testing.md#fixing-broken-visual-tests-inside-a-pr)
- [x] Ensure the changed/created components have not caused accessibility issues. [How to use accessibility plugin in storybook](https://github.com/toptal/picasso/blob/master/docs/contribution/accessibility.md).

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run all` - Run whole pipeline
- `@toptal-bot run build` - Check build
- `@toptal-bot run visual` - Run visual tests
- `@toptal-bot run deploy:documentation` - Deploy documentation
- `@toptal-bot run package:alpha-release` - Release alpha version

</details>

[SPC-1148]: https://toptal-core.atlassian.net/browse/SPC-1148

### CODE REVIEW

**Comment by s0ber** [comment]
@toptal-bot run package:alpha-release

**Comment by s0ber** [comment]
@toptal-bot run package:alpha-release

**Review by bogdanalexe90 — APPROVED** [review]

**Review by inemiro — APPROVED** [review]
Epic!

File: packages/picasso-lab/src/TypographyOverflow/TypographyOverflow.tsx [hunk 1]
```diff
@@ -41,6 +63,8 @@ export const TypographyOverflow = (props: Props) => {
     ...rest
   } = props
 
+  const [initialLines] = useState(lines)
+  const useStyles = useLinesStyles(initialLines)
```
Comment (OleksandrNechai) [inline packages/picasso-lab/src/TypographyOverflow/TypographyOverflow.tsx:67]: Do we ignore all further changes of the `lines` prop?

File: packages/picasso-lab/src/TypographyOverflow/TypographyOverflow.tsx [hunk 2]
```diff
@@ -23,9 +23,31 @@ export interface Props extends BaseProps, TypographyProps {
...
+// Naive approach with function rules that was used previously (https://cssinjs.org/jss-plugin-rule-value-function)
+// have shown a very poor performance because a new style tag was spawned for any new instance of TypographyOverflow component.
+// It resulted in massive UI freezes because TypographyOverflow is used a lot in many different places.
+// Please take a look at this PR for the explanation, usecases and the benchmarks
+// https://github.com/toptal/staff-portal/pull/3419
+const useLinesStyles = (lines: number) => {
+  let useStyles = lineStyles.get(lines)
+
+  if (useStyles) {
+    return useStyles
```
Comment (OleksandrNechai) [inline packages/picasso-lab/src/TypographyOverflow/TypographyOverflow.tsx:38]: I am not sure why this happens, maybe due to this statement, but when I testing multiline and then open the code editor I see this error

image

File: packages/picasso-lab/src/TypographyOverflow/TypographyOverflow.tsx [hunk 3]
```diff
@@ -23,9 +23,31 @@ export interface Props extends BaseProps, TypographyProps {
   disableTooltip?: boolean
 }
 
-const useStyles = makeStyles<Theme, Props>(styles, {
-  name: 'TypographyOverflow'
-})
+const lineStyles = new Map<number, ReturnType<typeof makeStyles>>()
```
Comment (OleksandrNechai) [inline packages/picasso-lab/src/TypographyOverflow/TypographyOverflow.tsx:26]: Why do we need this map if, after the first render in a given solution, the number of lines do not change?

Comment (s0ber) [reply packages/picasso-lab/src/TypographyOverflow/TypographyOverflow.tsx:26]: This is an map that holds style rules for different TypographyOverflow components that use different initial `lines` values. So all TypographyOverflow components that use 1 line will use a single style rules set. If a new component will appear that needs 2 lines - the new style rules set will be created and then reused for all other new instances.

File: packages/picasso-lab/src/TypographyOverflow/TypographyOverflow.tsx [hunk 4]
```diff
@@ -23,9 +23,31 @@ export interface Props extends BaseProps, TypographyProps {
...
+
+// This function creates separate styles object for any particular number of TypographyOverflow lines.
+// Naive approach with function rules that was used previously (https://cssinjs.org/jss-plugin-rule-value-function)
+// have shown a very poor performance because a new style tag was spawned for any new instance of TypographyOverflow component.
+// It resulted in massive UI freezes because TypographyOverflow is used a lot in many different places.
+// Please take a look at this PR for the explanation, usecases and the benchmarks
+// https://github.com/toptal/staff-portal/pull/3419
+const useLinesStyles = (lines: number) => {
```
Comment (OleksandrNechai) [inline packages/picasso-lab/src/TypographyOverflow/TypographyOverflow.tsx:34]: I fail to see why is this a hook? :-)

File: packages/picasso-lab/src/TypographyOverflow/TypographyOverflow.tsx (hunk 3 above)
Comment (OleksandrNechai) [reply packages/picasso-lab/src/TypographyOverflow/TypographyOverflow.tsx:26]: Oh, we reuse it between different instances of the component?

File: packages/picasso-lab/src/TypographyOverflow/TypographyOverflow.tsx (hunk 4 above)
Comment (s0ber) [reply packages/picasso-lab/src/TypographyOverflow/TypographyOverflow.tsx:34]: This is a function that creates a hook dynamically, not sure how to call it better :D
Will think about better name. Good catch 👍

**Review by elviocb — APPROVED** [review]
Great job @s0ber !!! 🚀

**Comment by s0ber** [comment]
@toptal-bot run package:alpha-release

File: packages/picasso-lab/src/TypographyOverflow/TypographyOverflow.tsx [hunk 5]
```diff
@@ -48,6 +49,16 @@ export const TypographyOverflow = (props: Props) => {
   const isTooltipRendered = isTooltipActive || isTooltipAnimating
   const isMultiline = lines > 1
 
+  // We are paying a very high price when using dynamic JSS rules
+  // for a component that is used on a very large scale.
+  // It was causing a major UI freezes and unnecessary style recalculations,
+  // that's why we decided to go with inline styles:
+  // https://github.com/toptal/picasso/pull/2110
+  const extendedStyle = useMemo(
```
Comment (OleksandrNechai) [inline packages/picasso-lab/src/TypographyOverflow/TypographyOverflow.tsx:57]: I believe `useMemo` is much more expensive than simply merging objects. I would remove it for code clarity.

**Review by OleksandrNechai — APPROVED** [review]

**Review by bogdanalexe90 — APPROVED** [review]

**Pull request merged**